### PR TITLE
Only set WCS observer if map frame has observer attribute

### PR DESCRIPTION
**Pull request by dstansby**
TL;DR: I think for a Stonyhurst frame, observer information shouldn't be set on a WCS.

Story time!

As part of https://github.com/sunpy/sunpy/pull/5244, warnings will be raised if plotting when `map.wcs != axes.wcs`. More precisely, this comparison is done with the line `axes.wcs.wcs.compare(self.wcs.wcs, tolerance=0.01)`. This will compare every WCS keyword, and evaluate true iff all keywords are identical to within an absolute tolerance of 0.01.

In the `reprojection_aia_euvi_mosaic.py` example, the following happens:

1. Create a FITS header in a Stonyhurst frame using `header_helper`. Because the Stonyhurst frame does (in general) not have an observer property, no observer keywords are set in the header. I shall call the WCS from this `wcs_header`.
2. Using `reproject`, create new maps from this header. I shall call the WCS from this `wcs_map`.
3. Create an axes with `wcs_header`
4. Plot the map on this axes

However, `map.wcs` sees there is no observer information in the header, and falls back to assuming the observer is the Earth, and sets the observer keywords in `wcs_map`. This leads to `wcs_header.compare(map_wcs, tolerance=0.01)` failing.

I *think* the solution to this is to only set the observer information in `GenericMap.wcs` if the coordinate frame of the map has an observer attribute. @ayshih should check this.

### CODE REVIEW

**Review by ayshih — COMMENTED** [review]
I'd like to slightly expand the scope of this PR.  Ideally, `GenericMap.observer_coordinate` should really be returning `None` if the map's coordinate frame is not an observer-based frame.  However, it's a bit tricky to fix in our current implementation because `GenericMap.observer_coordinate` is used as if it were a low-level property: `GenericMap.wcs` calls `GenericMap.observer_coordinate` (and `GenericMap.coordinate_frame` calls `GenericMap.wcs`).

I suggest:

- Rename the existing `observer_coordinate` property to some private name (and have the `wcs` property continue to use it)
- Create a new, high-level `observer_coordinate` property, which would simply be `getattr(self.coordinate_frame, 'observer', None)`

Thoughts?

**Comment by dstansby** [comment]
I had a bit of a think about the property call stack, and in the spirit of https://github.com/sunpy/sunpy/pull/5199 it sounds like we should be aiming for a world where

1. Map properties are drawn directly from the metadata
2. The `.wcs` should be constructed from these map properties (not directly from the metadata)

`.observer_coordinate` currently abides by 1., so I am reluctant to change it to being drawn from `.wcs` (via. `.coordinate_frame`). The issue is it defaults out to `get_earth()` instead of `None`, so by inspecting `.observer_coordinate` there is no way `.wcs` can currently know whether to set the `.observer_coordinate` or not.

I think the long term solution is therefore to make `.observer_coordinate` return `None` instead of `get_earth()`, probably after a deprecation period to avoid users tripping up.

**Comment by Cadair** [comment]
I agree with what @dstansby said, I would rather keep the properties as the canonical metadata extraction point.

We could make the observer return conditional on the frame type though, i.e. return `None` by default if the frame is Stonyhurst?

**Comment by dstansby** [comment]
A slight aside, but relevant for this discussion: given `HeliographicCarrington` has an `observer` to take into account light travel times, why doesn't `HeliographicStonyhurst`?

**Comment by ayshih** [comment]
> I agree with what @dstansby said, I would rather keep the properties as the canonical metadata extraction point.

They're _all_ properties.  How do you tell which properties are the "canonical" ones, and which properties can depend on other properties?  That is, why does `.observer_coordinate` have to be property to draw directly from metadata?

> I think the long term solution is therefore to make `.observer_coordinate` return `None` instead of `get_earth()`, probably after a deprecation period to avoid users tripping up.

xref #3387

**Comment by ayshih** [comment]
> A slight aside, but relevant for this discussion: given `HeliographicCarrington` has an `observer` to take into account light travel times, why doesn't `HeliographicStonyhurst`?

Reasons™

`HeliographicCarrington` (HGC) is intended for features on the surface of the Sun.  In that situation, the light travel time is well defined, as is the motion of the solar feature over the light travel time (albeit as a rigid rotator instead of a differential rotator).  This allows for the coordinate-feature correspondence to be maintained for observers at different Sun-observer distances.

`HeliographicStonyhurst` (HGS) is a general coordinate system, used for the location of any type of object anywhere in the heliosphere (or beyond), with no attempt to correct for light travel time.  This means that it is technically inaccurate to overlay HGS maps of the solar surface from two different observers, but the inaccuracy will be small if their Sun-observer distances are very similar (e.g., ~1 AU).  When combining AIA images with EUI images, it'd better to reproject to HGC.

**Comment by dstansby** [comment]
I'm going to put aside large scale design discussion for now, and write down the implementation loop currently spinning in my head...

- To know if a coordinate frame has an observer, we want to do `hasattr(frame, 'observer')`
- To get `frame` we use `astropy.wcs.utils.wcs_to_celestial_frame(wcs)`
- which means we need to construct `wcs`
- which means we need to know if the map coordinate frame has an observer (so we can set the observer keywords on the `wcs`)

If anyone can suggest where this vicious cycle can be disrupted I would be very grateful (and apologies if there is already a solution above somewhere)

**Comment by ayshih** [comment]
As I was fixing a different bug with Stonyhurst heliographic maps, I realized it was very related to the bug driving this PR.  So, I built PR #5478 off of this one.

**Comment by dstansby** [comment]
I've kind of lost track of this PR, so if you want to take over (or just make it part of another PR) then that would be very welcome.

**Comment by ayshih** [comment]
Merged in with #5478
